### ISSUE TITLE
Reduce code duplication between signature and code files

### ISSUE THREAD

**Issue by ivaylo5ev** [fsharp/fslang-suggestions#687]
I propose we make it so that certain type declarations from a signature file (.fsi) get compiled from the signature as if they were present in the respective .fs code file. This should apply to particular scenarios where there is no difference in the code between the signature type definition and the actual code file.

Currently, there are cases where some code duplication is required between the signature file (fsi) and its respective code file (fs). Below are some typical examples.

**DiscriminatedUnion.fsi**

    type ApplicationError =
        | RuntimeError of e:exn
        | InvalidArguments
        | SomethingElse

**InterfaceType.fsi**

    type [<Interface>] ISomeInterface =
        abstract member DoSomething: arg:obj -> unit

**RecordType.fsi**

    type EmailAddr = string
    type Person { Name:string; Email:EmailAddr }

The type alias above would also qualify.

The existing way of approaching this problem in F# is to duplicate the code from the sgnature in the respective code file. 

**Note**  

In the above examples I have chosen simple-enough cases where the code in both the signature and source file is **exactly** the same. This proposal should apply to such cases. However, I am aware of situations where the signature file may differ from the implementation, for instance:

**DiscreteDiscriminatedUnion.fsi**

    type ApplicationError 

Here the signature is deliberately hiding the DU type cases, thus the fs file is more elaborate. For cases like that, the current compiler behavior should be **enforced**. 

Other cases might involve putting extra attributes to the types in the code file (like `CustomEquality` and etc), which need not be mentioned in the signature. When a user type is provided following the signature, the compiler should use that and must not attempt to compile the respective type from the signature file.

So, in general, the suggestion would enable skipping certain type definitions from the code files, by automatically generating the type from the signature during compilation. In case the developer has defined an implementing type, the auto-generation is skipped in favor of the user code. If a type implementation is not present, and the type's exact definition cannot be inferred by the compiler (like in the discrete DU type above), then an error should be presented forcing the user to provide an implementation. This should be well in par with how the compiler works now.

## Pros and Cons 

The advantages of making this adjustment to F# are:
 * Interfaces and **simple enough** DU types can be defined only once in the signature file
 * Code changes happen in one place
 * The user code is less verbose for simple type definitions

The disadvantages of making this adjustment to F# are: 
 * The F# compiler does not seem to care where a type from a .fsi file is defined. 
  There may not even exist a respective fs file, but the signature may still be implemented across different fs files. I am not familiar enough with the compiler to tell whether this could pose a problem-- for example could the user be required to have a respective *.fs file along (one with the same name before the extension ans the signature), which may be inconvenient.

## Extra information

Estimated cost (XS, S, M, L, XL, XXL):  
  Maybe L or XL (my knowledge of the compiler is not good to give an accurate estimate here)

Related suggestions: (put links to related suggestions here)
  I've not spotted similar ones so far but will be happy to know if such exist.

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [X] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [X] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [X] This is not a breaking change to the F# language design
I believe this will not be a breaking change, because existing code will conform to the rules of this suggestion and should still compile in the same manner.

* [x] I or my company would be willing to help implement and/or test this. 
I am willing to assist with implementing this feature, if approved. My knowledge of the F# compiler at this time is not really great, so I will probably be taking my time getting familiar along the way.

**Comment by ivaylo5ev** [comment]
An additional set of candidate type definitions for this proposal:
* type aliases
    type sometype = string
* record types
    type MyRec = { Member:type ... }

**Comment by kerams** [comment]
Yeah, interfaces and DUs, records, enums without extra public members should definitely be allowed to live solely in signature files. Another issue with the status quo is the fact that xmldoc can easily get out of sync when you want both definitions just to stay identical.

Would be really handy in the compiler itself, I can see it saving hundreds of lines of code.

**Comment by cartermp** [comment]
The thing that's troubling about this for me is how non-uniform it is. And if F# supports defining things like default interface members (it can consume but not define today) then that's yet another place where "it depends" on whether or not you can use a shortcut like this.

**Comment by smoothdeveloper** [comment]
@kerams, regarding the compiler code itself, maybe it is worth going the route of moving those types in plain .fs files?

I am nut sure it would be more work than if this suggestion existed and doing the refactor you propose; although this route, one issue is about nested modules and splitting among translation units, which wouldn't work, IIRC.

I don't use the signature files much, but I relates to what @cartermp says about reducing uniformity of how the signature files work right now, and how signature syntax will evolve; it could introduce ambiguity with other suggestions.

Is there such an invariant about: "if one removes .fsi files from projects, it still compiles"? This suggestion would disrupt it.
